enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+
+## Motivation
+
+Reduce administrative overhead to run a cluster and ability to respond to failures of machines.
+
+### Goals
+
+- Providing a default remediation strategy for public clouds - Machine deletion.
+
+- Allowing to define different unhealthy criterias for different pools of machines.
+
+- Allowing new and custom remediation strategies developed outside the machine-healthcheck controller, e.g baremetal reboot.
+
+- Providing administrator-defined short-circuiting of automated remediation.
+
+
+### Non-Goals
+
+- Coordinating individual remediation systems. This must be driven by each system however they see fit, e.g. by annotating affected nodes with state and transition time.
+
+- Recording long-term stable history of all health-check failures or remediations.
+
+## Proposal
+
+- As a customer I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated automatically.
+
+- As a cluster admin I want to define my unhealthy criteria as a list of node conditions https://github.com/kubernetes/node-problem-detector.
+
+- As a cluster admin I want to define different unhealthy criteria for pools of machines targeted for different workloads.
+
+- As a cluster admin I want to define different remediation strategies for different cloud environments. E.g deletion or reboot.
+
+- As a cluster admin I want to skip remediation to optionally prevent automated remediation if some threshold of nodes are unhealthy, because this situation may require more careful intervention. E.g a large number of nodes in a single zone are down because it's most likely a networking issue.
+
+Machine health checking is an integration point between node problem detection tooling and remediation strategies.
+
+We will provide a default remediation strategy - delete. It will delete a machine when the backed node meets the unhealthy criteria defined by the user. The upper level machine controller, e.g machineSet will then reconcile towards satisfying the expected number of replicas and will create a new one.
+
+### Implementation Details
+
+#### MachineHealthCheck CRD:
+- Enable watching a pool of machines (based on a label selector).
+- Enable defining an unhealthy criteria (based on a list of node conditions).
+- Enable choosing a remediation strategy.
+- Enable setting a threshold for the number of unhealthy nodes. If the current number is at or above this threshold no further remediation 
+
+
+E.g I want my worker machines in us-east-1a to be remediated when the backed node has `ready=false` condition for more than 10m.
+
+```yaml
+apiVersion: "healthchecking.openshift.io/v1beta1"
+kind: "MachineHealthCheck"
+metadata:
+  name: workers-us-east-1a
+  namespace: openshift-machine-api
+spec:
+  selector:
+    role: worker
+    zone: us-east-1a
+  remediationStrategy: delete
+  unhealthyConditions:
+    - name: Ready
+      status: false
+      timeout: 10m
+  maxUnavailable: 5
```

Comment:
I expect percentages to work here?  With the same behavior as deployments /pdbs in terms of constraints?